storage/filestore/dir.go

```diff
@@ -238,6 +348,136 @@ func (dir *Dir) GarbageCollect(ctx context.Context) (err error) {
+// format V1 or greater, in the given namespace. If doForEach returns a non-nil error,
+// ForAllKeysInNamespace will stop iterating and return the error immediately.
+func (dir *Dir) ForAllKeysInNamespace(ctx context.Context, namespace []byte, doForEach func(storage.StoredBlobAccess) error) (err error) {
+	namespaceDir := pathEncoding.EncodeToString(namespace)
+	nsDir := filepath.Join(dir.blobsdir(), namespaceDir)
+	openDir, err := os.Open(nsDir)
+	if err != nil {
+		if os.IsNotExist(err) {
+			// job accomplished: there are no blobs in this namespace!
+			return nil
+		}
+		return err
+	}
+	defer func() { err = errs.Combine(err, openDir.Close()) }()
+	for {
+		// check for context done both before and after our readdir() call
+		if err := ctx.Err(); err != nil {
+			return err
+		}
+		subdirNames, err := openDir.Readdirnames(nameBatchSize)
+		if err != nil && err != io.EOF {
+			return err
+		}
+		if os.IsNotExist(err) || len(subdirNames) == 0 {
+			return nil
+		}
+		if err := ctx.Err(); err != nil {
+			return err
+		}
+		for _, keyPrefix := range subdirNames {
+			if len(keyPrefix) != 2 {
+				// just an invalid subdir; could be garbage of many kinds. probably
+				// don't need to pass on this error
+				continue
+			}
+			err := dir.forAllKeysInNamespaceWithPrefix(ctx, namespace, nsDir, keyPrefix, doForEach)
+			if err != nil {
+				return err
+			}
+		}
+	}
+}
+
+func (dir *Dir) forAllKeysInNamespaceWithPrefix(ctx context.Context, namespace []byte, nsDir, keyPrefix string, doForEach func(storage.StoredBlobAccess) error) (err error) {
+	keyDir := filepath.Join(nsDir, keyPrefix)
+	openDir, err := os.Open(keyDir)
+	if err != nil {
+		return err
+	}
+	defer func() { err = errs.Combine(err, openDir.Close()) }()
+	for {
+		// check for context done both before and after our readdir() call
+		if err := ctx.Err(); err != nil {
+			return err
+		}
+		keyInfos, err := openDir.Readdir(nameBatchSize)
+		if err != nil && err != io.EOF {
+			return err
+		}
+		if os.IsNotExist(err) || len(keyInfos) == 0 {
+			return nil
+		}
+		if err := ctx.Err(); err != nil {
+			return err
+		}
+		for _, keyInfo := range keyInfos {
+			if keyInfo.Mode().IsDir() {
+				continue
+			}
+			blobFileName := keyInfo.Name()
+			encodedKey := keyPrefix + blobFileName
+			formatVer := storage.FormatV0
+			if strings.HasSuffix(blobFileName, v1PieceFileSuffix) {
+				formatVer = storage.FormatV1
+				encodedKey = encodedKey[0 : len(encodedKey)-len(v1PieceFileSuffix)]
+			}
+			key, err := pathEncoding.DecodeString(encodedKey)
+			if err != nil {
+				continue
+			}
+			ref := storage.BlobRef{
+				Namespace: namespace,
+				Key:       key,
+			}
+			fullPath := filepath.Join(keyDir, blobFileName)
+			err = doForEach(newStoredBlobAccess(ref, fullPath, keyInfo, formatVer))
+			if err != nil {
+				return err
+			}
+			// also check for context done between every doForEach callback.
+			if err := ctx.Err(); err != nil {
```

Comment:
For all of these checks, why not just do

```
if ctx.Err() != nil {
    return ctx.Err()
}
```

Seems it would save one allocation. I don't see an issue with a race of any kind, since according to the context documentation:

`        // After Err returns a non-nil error, successive calls to Err return the same error.`

Reply:
There probably is no race. However, using a variable like this reads cleaner to me, doesn't affect performance, and doesn't send me off to the Go documentation to make sure that context.Context objects are guaranteed not to change their error return values :)

Reply:
Actually, calling `ctx.Err()` twice probably does slow things down a tiny bit, as the compiler probably can't tell that it's guaranteed to get the same value both times, so the call and the processing can't be optimized out the way this variable can.